pkg/iosrc/uri.go

```diff
@@ -0,0 +1,87 @@
+package iosrc
+
+import (
+	"errors"
+	"net/url"
+)
+
+type URI struct {
+	url.URL
+}
+
+const (
+	stdin  = "stdio:///stdin"
+	stdout = "stdio:///stdout"
+	stderr = "stdio:///stderr"
+)
+
+// ParseURI parses the path using `url.Parse`. If the provided uri does not
+// contain a scheme, the scheme will set to file. Relative paths will be
+// treated as files and resolved as absolute paths using filepath.Abs.
+// path is an empty, Scheme is set to file.
+func ParseURI(path string) (URI, error) {
+	// First resolve stdio keywords in to fully-formed uri.
+	path = stdio(path)
+	var err error
+	path, err = normalizeFilepaths(path)
+	if err != nil {
+		return URI{}, err
+	}
+	u, err := url.Parse(path)
+	if err != nil {
+		return URI{}, err
+	}
+	return URI{*u}, nil
+}
+
+func stdio(path string) string {
+	switch path {
+	default:
+		return path
+	case "stdout":
+		return stdout
+	case "stdin":
+		return stdin
+	case "stderr":
+		return stderr
+	}
+}
+
+func (p URI) AppendPath(elem ...string) URI {
+	for _, el := range elem {
+		p.Path = p.Path + "/" + el
+	}
```

Comment:
```suggestion
	p.Path = path.Join(append([]string{p.Path}, elem...))
```

Reply:
path.Join also does a call to path.Clean that was messing things up for me. Don't want that.

Reply:
In that case, maybe `p.Path = strings.Join(append([]string{p.Path}, elem...), "/")`. Or maybe not.